lib/block_view.go

```diff
@@ -3593,6 +3648,139 @@ func (bav *UtxoView) _connectTransaction(txn *MsgDeSoTxn, txHash *BlockHash,
+	gp := bav.GetCurrentGlobalParamsEntry()
+
+	failingTransactionRate := uint256.NewInt().SetUint64(gp.FailingTransactionBMFRateBasisPoints)
+	failingTransactionFee := uint256.NewInt().SetUint64(txn.TxnFeeNanos)
+	basisPointsAsUint256 := uint256.NewInt().SetUint64(10000)
+
+	effectiveFeeU256 := failingTransactionRate.Mul(failingTransactionRate, failingTransactionFee)
+	effectiveFeeU256.Div(effectiveFeeU256, basisPointsAsUint256)
+	// We should never overflow on the effective fee, since FailingTransactionBMFRateBasisPoints is <= 10000.
+	// But if for some magical reason we do, we set the effective fee to the max uint64. We don't error, and
+	// instead let _spendBalance handle the overflow.
+	if !effectiveFeeU256.IsUint64() {
+		effectiveFeeU256.SetUint64(math.MaxUint64)
+	}
+	effectiveFee := effectiveFeeU256.Uint64()
+	// If the effective fee is less than the minimum network fee, we set it to the minimum network fee.
+	if effectiveFee < gp.MinimumNetworkFeeNanosPerKB {
+		effectiveFee = gp.MinimumNetworkFeeNanosPerKB
+	}
+	burnFee, utilityFee := computeBMF(effectiveFee)
+
+	var utxoOps []*UtxoOperation
+	// When spending balances, we need to check for immature block rewards. Since we don't have
+	// the block rewards yet for the current block, we subtract one from the current block height
+	// when spending balances.
+	feeUtxoOp, err := bav._spendBalance(effectiveFee, txn.PublicKey, blockHeight-1)
+	if err != nil {
+		return nil, 0, 0, errors.Wrapf(err, "_connectFailingTransaction: Problem "+
+			"spending balance")
+	}
+	utxoOps = append(utxoOps, feeUtxoOp)
+
+	// If verifySignatures is passed, we check transaction signature.
+	if verifySignatures {
+		if err := bav._verifyTxnSignature(txn, blockHeight); err != nil {
+			return nil, 0, 0, errors.Wrapf(err, "_connectFailingTransaction: Problem "+
+				"verifying signature")
+		}
+	}
+
+	return utxoOps, burnFee, utilityFee, nil
+}
+
+// computeBMF computes the burn fee and the utility fee for a given fee. The acronym stands for Burn Maximizing Fee, which
+// entails that the burn function is designed to maximize the amount of DESO burned, while providing the minimal viable
+// utility fee to the block producer. This is so that block producers have no advantage over other network participants
+// in accruing DESO. The utility fee that block producers get from transaction fees is just enough so that it's economically
+// advantageous to include as many transactions as possible in blocks. Reliably doing so for an extended period of time
+// could accumulate to some DESO from transaction fees.
+// The utility fee is computed by taking a binary logarithm of the fee, and the remainder is burned. BMF(fee) computes:
+//
+//	burnFee := fee - log_2(fee), utilityFee := log_2(fee).
+func computeBMF(fee uint64) (_burnFee uint64, _utilityFee uint64) {
+	// Compute the utility fee as log_2(fee). We can find it by taking the bit length of fee.
+	utilityFee := uint64(bits.Len64(fee))
```

Comment:
I wouldn't be clever here. It's better to just use BigFloatLog or a standard function because 1) it's easier to read/verify and 2) it's easier to change it in the future.